corehq/apps/users/util.py

```diff
@@ -245,6 +245,7 @@ def update_latest_builds(user, app_id, date, version):
             user.reporting_metadata.last_builds.append(last_build)
         last_build.build_version = version
         last_build.app_id = app_id
+        last_build.build_profile_id = build_profile_id
```

Comment:
This will only get updated the next time the user updates the app unless you also check the profile ID in `_last_build_needs_update`

Reply:
thanks for sharing that,
I looked at that method and i thought it would never return `False`, because of the way it compares the date

https://github.com/dimagi/commcare-hq/blob/afbe8c8f6cc7b57ce325f3512fd886117550bce8/corehq/apps/users/util.py#L227-L232

Reply:
pretty sure that will return False if `last_build.build_version_date` is not falsy and `build_date <= last_build.build_version_date`

Reply:
that is correct.

Apologies if i am understanding this incorrectly. Just going to write down what I understand.

In case of heartbeat the value of date [passed to this function](https://github.com/dimagi/commcare-hq/blob/153bce361ae7d7f0d1e7c0e58cd7f7479cc895a0/corehq/apps/ota/views.py#L342) is `datetime.utcnow()` which should ideally be greater than `last_build.build_version_date`. The scenario you are highlighting is possible in case this is called for [form submission](https://github.com/dimagi/commcare-hq/blob/3ac2b4e7e6bfe746bd973391999de4b5b879bda3/corehq/ex-submodules/pillowtop/processors/form.py#L160) or [synclog](https://github.com/dimagi/commcare-hq/blob/3ac2b4e7e6bfe746bd973391999de4b5b879bda3/corehq/pillows/synclog.py#L98) in case they are not received in order of occurrence on mobile.
the build profile id is set only via heartbeat, so i think this would update build profile id for each heartbeat request.

Reply:
Interesting, I was not looking at how this was called an going only on the variable names which are clearly misleading. Given what these dates are (form.received_on, synclog.date, heartbeat request time) this method hardly makes any sense. If the date were in fact the `build_date` then it makes much more sense. As it stands it appears that the function will almost ALWAYS return True unless it is being called with an old form / synclog.

That could also explain why this put so much load on couch. Not sure what the best approach to fixing that is but it feels like we should either be comparing the date of the app build or else just the app version.

Reply:
Since the build version is already available I think it makes most sense to check that `version > last_build.build_version` instead of using the date.

Reply:
thanks for weighing in @snopoke 
I think thats a fair point from the load perspective and should be improved. Can we do that in a separate PR?